ovgs.proto

```diff
@@ -1,175 +1,241 @@
-    rpc viewGroup(viewGroupRequest) returns (viewGroupResponse);
-    // deleteGroup deletes a named group. Will refuse to do so if there are children groups.
-    rpc deleteGroup(deleteGroupRequest) returns (deleteGroupResponse);
-    // assignUserRole assigns a role to a user in a named group.
-    rpc assignUserRole(assignUserRoleRequest) returns (assignUserRoleResponse);
-    // removeUserRole removes roles of a user in a named group.
-    rpc removeUserRole(removeUserRoleRequest) returns (removeUserRoleResponse);
-    // viewUserRole returns the roles of a user across the groups that they belong to.
-    rpc viewUserRole(viewUserRoleRequest) returns (viewUserRoleResponse);
-    // assignSerial assigns the serial to that group.
-    rpc assignSerial(assignSerialRequest) returns (assignSerialResponse);
-    // unassignSerial removes the serial from the group.
-    rpc unassignSerial(unassignSerialRequest) returns (unassignSerialResponse);
-    // viewSerial returns all the facts about the serial number.
-    rpc viewSerial(viewSerialRequest) returns (viewSerialResponse);
-    // setDomainCert sets the values for the certificate.
-    rpc setDomainCert(setDomainCertRequest) returns (setDomainCertResponse);
-    // readDomainCert reveals the details of the certificate.
-    rpc readDomainCert(readDomainCertRequest) returns (readDomainCertResponse);
-    // deleteDomainCert deletes the cert from the database.
-    rpc deleteDomainCert(deleteDomainCertRequest) returns (deleteDomainCertResponse);
-    // ownershipVoucher issues ownership voucher and returns TPM public key for the serial number
-    rpc ownershipVoucher(ownershipVoucherRequest) returns (ownershipVoucherResponse);
+  // CreateGroup creates a group as a child of an existing group.
+  // Errors will be returned:
+  // ALREADY_EXISTS if group already exists
+  // PERMISSION_DENIED if the user doesn't have access to the parent group
+  rpc CreateGroup(CreateGroupRequest) returns (CreateGroupResponse);
+  // DeleteGroup deletes a named group.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // FAILED_PRECONDITION if the parent is set but does not exist
+  // PERMISSION_DENIED if user doesn't have access to parent group
+  rpc DeleteGroup(DeleteGroupRequest) returns (DeleteGroupResponse);
+  // GetGroup returns the domain-certs (keyed by id), serials,
+  // and user/role mappings for that group and all the children.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc GetGroup(GetGroupRequest) returns (GetGroupResponse);
+
+  // CreateUserRole assigns a role to a user in a named group.
+  // Errors will be returned:
+  // ALREADY_EXISTS if user already exists in the group
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc CreateUserRole(AssignUserRoleRequest) returns (AssignUserRoleResponse);
```

Comment:
Can we stick to AssignUserRole (and not call it CreateUserRole)?

There are only 2-3 valid roles. Users can be assigned any of these roles. CreateUserRole sounds like we're letting people create roles, rather than assigning predefined roles to users.

Reply:
qq: CreateUserRequest/AddUser and DeleteUser(DeleteUserRequest).

Is the suggestion that we allow creation of new users/service accounts and assign them roles? or is it to assign a role to an existing user/service account?

The latter is what currently implements. Up until now, the interface does not talk about how users/service account are created, just that they do, and the AssignUserRole rpc allowed giving them a role. There are other endpoints to create users/service accounts (We shared them with heas@goog a while ago). It would be useful to nail this down, since the former would require some work.

Reply:
The ask isn't to allow creation of user and service accounts. Just mark the externalities as such. Added more clarifications.

Reply:
`as it looks like users are always tied to groups and they are "created" / "deleted" in the context of that group?` They're not. Users/service accounts can exist (defined via some api, which isn't part of the ovgs service) in by themselves. Roles are assigned to users in the context of some group. 

`we should document the user / service account definitions are being done via another api and are available to this api` I added the following comment above the User type -

// Creation of user accounts and service accounts (in a given org, denoted by
// it's org_id) is external to the ovgs service. Accounts may be created by
// any mechanism that a service implementor may offer. This service only deals
// with the assignment of roles to these existing user/service accounts in the context
// of a group.

Reply:
@sulrich